### DISCUSSION TITLE
Unexpected cache miss

### DISCUSSION THREAD

**Discussion by chrismeyers** [vercel/vercel#9280] [Help] [answered]
I've deployed a React + Vite website to Vercel and have noticed some weird behavior surrounding the `x-vercel-cache` header.  No header configuration has been changed in `vercel.json`.

### Steps to reproduce:
1. Go to https://chrismeyers.net and open the network tab in dev tools
2. Hard reload the page, all items served from Vercel have a response code of 200 and `x-vercel-cache: HIT` header
3. Reload the page (not hard reload), all items served from Vercel have a response code of 304 but only the HTML has a `x-vercel-cache: HIT` header
<details>
    <summary>curl commands</summary>

HIT:
```
curl -I 'https://chrismeyers.net/' \
  -H 'authority: chrismeyers.net' \
  -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'cache-control: max-age=0' \
  -H 'if-none-match: "6c078ba3740d71177502b6d9d2abc2fd"' \
  -H 'sec-ch-ua: "Not_A Brand";v="99", "Google Chrome";v="109", "Chromium";v="109"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-dest: document' \
  -H 'sec-fetch-mode: navigate' \
  -H 'sec-fetch-site: same-origin' \
  -H 'sec-fetch-user: ?1' \
  -H 'upgrade-insecure-requests: 1' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36' \
  --compressed
```

MISS:
```
curl -I 'https://chrismeyers.net/assets/index-72d38370.js' \
  -H 'authority: chrismeyers.net' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'if-none-match: W/"0979e20ccd9e1bbf7256a278f8b5ad7d"' \
  -H 'origin: https://chrismeyers.net' \
  -H 'referer: https://chrismeyers.net/' \
  -H 'sec-ch-ua: "Not_A Brand";v="99", "Google Chrome";v="109", "Chromium";v="109"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-dest: script' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36' \
  --compressed
```
</details>

### Expected:
I expect every item served from Vercel in step 3 to have `x-vercel-cache: HIT`.  Is there a reason why the JS, CSS, and favicon have a `x-vercel-cache: MISS` header in this case?

**Comment by amyegan** [comment]
By default the `Cache-Control` header value is `public, max-age=0, must-revalidate` which tells the Edge Network and the browser not to cache. 

Static files, such as your HTML, do not change until a new deployment invalidates the cache. The default `Cache-Control` header prevents browsers from caching the file locally, but the files are automatically cached at the edge after the first request. That way users get the latest file from the Global CDN immediately after deploying without the risk of local cache delaying new content.

Serverless Functions, on the other hand, are more dynamic. For the CDN to cache the response of a Serverless Function, you need to define the number of seconds a response should be cached using a customized [Cache-Control](https://vercel.com/docs/concepts/edge-network/headers#cache-control-header) header. Otherwise, the Function will be executed each time it is called, even if the response was a 304. 

When you see a `304` status with `x-vercel-cache: HIT`, that means the cached resource was used. But when you see a `304` status with `x-vercel-cache: MISS`, that means that the Serverless Function was executed.

More info:
- [Cache-Control Header](https://vercel.com/docs/concepts/edge-network/headers#cache-control-header)
- [Caching](https://vercel.com/docs/concepts/edge-network/caching#)
  - [Static Files Caching](https://vercel.com/docs/concepts/edge-network/caching#static-files-caching)
  - [Serverless Functions](https://vercel.com/docs/concepts/edge-network/caching#serverless-functions)

**Reply by chrismeyers** [reply]
That makes sense for serverless functions, but I would expect the JS and CSS to be served as static assets.  Is that not the case?

**Reply by amyegan** [reply]
CSS and JS are commonly static resources, but they may be server-side generated depending on the frameworks and settings for your app. I can't give any specific recommendations at this point since I don't know your code. 

If you are able to share your code (or a minimal reproduction) and some info about your project setting, then someone here might be able to review it to help you identify exactly what is happening in your case. 

Barring that, I can only make some guesses as to what is relevant to your situation. Here are some links that I hope will help:
- Vercel docs
  - [Edge Network Caching](https://vercel.com/docs/concepts/edge-network/caching#)
  - [How to Configure the Cache-Control Response Header in Vercel Projects](https://vercel.com/guides/how-to-configure-the-cache-control-response-header-in-vercel-projects)
  - [Caching responses from Edge Functions](https://vercel.com/docs/concepts/functions/edge-functions/edge-caching#)
  - [Caching responses from Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions/edge-caching#)
- Next.js docs
  - [Going to Production: Caching](https://nextjs.org/docs/going-to-production#caching)
  - [Caching with Server-Side Rendering (SSR)](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#caching-with-server-side-rendering-ssr)
  - [Built-In CSS Support](https://nextjs.org/docs/basic-features/built-in-css-support)
  - [Customizing PostCSS Config](https://nextjs.org/docs/advanced-features/customizing-postcss-config)

**Reply by chrismeyers** [reply]
Sure, the code in question is located here: https://github.com/chrismeyers/website

It's a pretty simple client side rendered SPA using React.  Vite is used to generate static production assets (one JS bundle and one CSS bundle).

Let me know if you need any more information, thanks for the help!

**Comment by chrismeyers** [comment, marked as answer]
I think something changed since I'm now seeing HTTP 304 responses with `x-vercel-cache: HIT` for all resources (HTML, JS, CSS, favicon) as expected
